drivers/counter/timer_esp32.c

```diff
@@ -0,0 +1,290 @@
+
+struct timer_isr_func_t {
+	timer_isr_t fn;
+	void *args;
+	timer_isr_handle_t timer_isr_handle;
+	timer_group_t isr_timer_group;
+};
+
+struct counter_obj_t {
+	timer_hal_context_t hal;
+	struct timer_isr_func_t timer_isr_fun;
+};
+
+struct counter_esp32_config {
+	struct counter_config_info counter_info;
+	timer_config_t config;
+	timer_group_t group;
+	timer_idx_t idx;
+
+	const struct {
+		int source;
+		int line;
+	} irq;
+
+	counter_irq_config_func_t irq_config_fn;
+};
+
+struct counter_esp32_data {
+	struct k_sem sem;
+	counter_alarm_callback_t callback;
+	uint32_t ticks;
+	void *user_data;
+};
+
+static struct counter_obj_t *p_timer_obj[TIMER_GROUP_MAX][TIMER_MAX] = { 0 };
+
+static inline void counter_esp32_sem_take(const struct device *dev)
+{
+	k_sem_take(&DEV_DATA(dev)->sem, K_FOREVER);
+}
+
+static inline void counter_esp32_sem_give(const struct device *dev)
+{
+	k_sem_give(&DEV_DATA(dev)->sem);
+}
+
+static int counter_esp32_init(const struct device *dev)
+{
+	const struct counter_esp32_config *cfg = DEV_CFG(dev);
+
+	k_sem_init(&DEV_DATA(dev)->sem, 1, 1);
+
+	if (TIMG(dev) == TIMER_GROUP_0) {
+		periph_module_enable(PERIPH_TIMG0_MODULE);
+	} else if (TIMG(dev) == TIMER_GROUP_1) {
+		periph_module_enable(PERIPH_TIMG1_MODULE);
+	} else {
+		return -ENOTSUP;
+	}
+
+	if (TIMX == NULL) {
+		TIMX = (struct counter_obj_t *)k_calloc(1, sizeof(struct counter_obj_t));
+		if (TIMX == NULL) {
+			LOG_ERR("TIMER driver malloc error");
+			return -ENOMEM;
+		}
+	}
+
+	counter_esp32_sem_take(dev);
+	timer_hal_init(&TIMX->hal, TIMG(dev), TIDX(dev));
+	timer_hal_intr_disable(&TIMX->hal);
+	timer_hal_clear_intr_status(&TIMX->hal);
+	timer_hal_set_auto_reload(&TIMX->hal, cfg->config.auto_reload);
+	timer_hal_set_divider(&TIMX->hal, cfg->config.divider);
+	timer_hal_set_counter_increase(&TIMX->hal, cfg->config.counter_dir);
+	timer_hal_set_alarm_enable(&TIMX->hal, cfg->config.alarm_en);
+	if (cfg->config.intr_type == TIMER_INTR_LEVEL) {
+		timer_hal_set_level_int_enable(&TIMX->hal, true);
+	}
+	timer_hal_set_counter_value(&TIMX->hal, INITIAL_COUNT);
+	timer_hal_set_counter_enable(&TIMX->hal, cfg->config.counter_en);
+	DEV_CFG(dev)->irq_config_fn(dev);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_start(const struct device *dev)
+{
+	counter_esp32_sem_take(dev);
+	timer_hal_set_counter_enable(&TIMX->hal, TIMER_START);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_stop(const struct device *dev)
+{
+	counter_esp32_sem_take(dev);
+	timer_hal_set_counter_enable(&TIMX->hal, TIMER_PAUSE);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_get_value(const struct device *dev, uint32_t *ticks)
+{
+	counter_esp32_sem_take(dev);
+	timer_hal_get_counter_value(&TIMX->hal, (uint64_t *)ticks);
+	counter_esp32_sem_give(dev);
+
+	return 0;
+}
+
+static int counter_esp32_set_alarm(const struct device *dev, uint8_t chan_id,
+				   const struct counter_alarm_cfg *alarm_cfg)
+{
```

Comment:
it's missing check against channel being busy.

Reply:
It is the intended behavior if I got the comment right. `chan_id` is not used in this timer implementation and is handled by `ARG_UNUSED` macro to avoid build-time warnings.